waku/v2/peermanager/peer_manager.go

```diff
@@ -241,6 +243,98 @@ func (pm *PeerManager) Start(ctx context.Context) {
 		go pm.peerEventLoop(ctx)
 	}
 	go pm.connectivityLoop(ctx)
+	go pm.peerStoreLoop(ctx)
+}
+
+func (pm *PeerManager) peerStoreLoop(ctx context.Context) {
+	t := time.NewTicker(prunePeerStoreInterval)
+	defer t.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-t.C:
+			pm.prunePeerStore()
+		}
+	}
+}
+
+func (pm *PeerManager) prunePeerStore() {
```

Comment:
I'm wondering if it's possible to do this,
- addPeer, check if meets maxPeers, if yes, return, otherwise add the peer.
- regular check and clear disconnected peers and meets the max retry.
- we may also need the max shards supported, each shard may have different maxPeers. 

This seems a bit easier to maintain.

Reply:
> addPeer, check if meets maxPeers, if yes, return, otherwise add the peer.

this was there before, but this is not proper as we don't know on what basis we are blocking new peer addition.

Reply:
> we may also need the max shards supported, each shard may have different maxPeers.

did not get this...why do we need max shards supported and maxPeers per shard? maxPeers indicates capacity in peerStore which should be for whole node. can you elaborate on any specific benefits you see with this approach.

Reply:
I see there is average calculation in your code, and this maxPeerPerShard may improve it and not bring too much burden to the whole node, also benefits scenarios like different shard needs different max peers for connection (this may not be true or needed).

Reply:
Hmm, maybe we can consider min peer per shard(but then again this may be hard in scenarios where a new shard is just coming up and only community owner's node is present)? what number should we consider for maxPeerPerShard ?

`scenarios like different shard needs different max peers for connection` 

why would this be a config? maybe not getting what you mean.

Reply:
For relay, the minPeerPerShard should be 4 to maintain network healthy, but lower can be accepted if really no available peers or poor connection situation. The maxPeerPerShard is more like a bandwidth consideration, could be around 6 by default, isn't this already set by libp2p pubsub protocol?

For other protocols, we may have different min and max peers for connection.

I may misunderstand the point of this PR, shouldn't peer store only clear dead connection?

maxShards and maxPeerPerShard (the sum of maxPeerPerProtocol in one shard) may provide an alternative way to maxPeers related logic. 

With more thinking, we don't need to configure different min/max peers for different shard until there is a strong requirement for this.

The current implementation LGTM, providing above feedback for discussion.

Reply:
`For relay, the minPeerPerShard should be 4 to maintain network healthy, but lower can be accepted if really no available peers or poor connection situation. The maxPeerPerShard is more like a bandwidth consideration, could be around 6 by default, isn't this already set by libp2p pubsub protocol`

i understand wrt minPeersPerShard, but maxPeerPerShard is where i assumed wrong as well. This is actual full-mesh peers and connected peers may not necessarily mean full-mesh peers. We need more connected peers in general so that gossip happens wrt messages in the network. otherwise it leads to unhealthy mesh. ref https://github.com/waku-org/go-waku/issues/1132


`For other protocols, we may have different min and max peers for connection.`

true, for light protocols this has to be determined but as of now i am thinking something like min of 2 per shard and max of 5-6. 

`maxShards and maxPeerPerShard (the sum of maxPeerPerProtocol in one shard) may provide an alternative way to maxPeers related logic.`

while implementing changes for lightClient i did realize that somewhere we need to link number of shards with maxPeerCount otherwise it may become tough. So, thanks for bringing this up. will look into it.

e.g: consider status communties where 2 common shards are present and a user is part of 20 communities which are on their own shards. in this case maxPeers should be `22*<per shard value>` .  Per shard value would be based on protocols mounted.